### ISSUE TITLE
How to change my URDF file to follow your convention

### ISSUE THREAD

**Issue by popi-mkx3** [chvmp/champ_setup_assistant#3]
Hello,

First of all, thank you for your work, this is good and well documented.
After I played around for a while with Champ and Spot, I tried to give it a try and generate the configuration for my own robot (popi).

I already have a URDF file, though it doesn't follow the assumptions you made (I created it following the modified Denavit-Hartemberg convention to fix my frames). It is still quite close, as you can see in the following screens : 

![Capture d’écran de 2020-10-09 19-21-11](https://user-images.githubusercontent.com/62474461/95615820-21369080-0a69-11eb-8de0-c2da8e726b18.png)

![Capture d’écran de 2020-10-09 19-21-38](https://user-images.githubusercontent.com/62474461/95615824-2267bd80-0a69-11eb-9529-4d9c6203685d.png)

![Capture d’écran de 2020-10-09 19-42-12](https://user-images.githubusercontent.com/62474461/95615825-2267bd80-0a69-11eb-87f7-8949e282fb2c.png)

The main difference I see is about the "zero" position, it was set with the legs fully bent. A friend of mines generated this URDF file using a solidworks to URDF exporter and defined the meshes origins and all the frames in the process. I am not really proficient with URDF, so I don't really know how I could change these settings directly in the URDF file. Is it even possible, or would I have to go through the exporter again to get a correct configuration ?

If you'd like to have a look at the URDF file, you can find it [here](https://github.com/popi-mkx3/popi_project/tree/master/popi_software/popi/popi_description)

With regards,
Lucas

**Comment by grassjelly** [comment]
Hey Lucas,

Thanks for your interest in CHAMP. Unfortunately, the assumptions must be strictly adhered for now. This is to generalize the IK engine across multiple robots. I took a quick glance on your URDF and I think I can help to make the changes. If you're okay I can push the generated config file as well in the official robots too.

**Comment by grassjelly** [comment]
Hey Lucas,

I did a copy of the xacro file on my forked version that makes Popi's URDF compatible with CHAMP. https://github.com/chvmp/popi_project. Let me know if you're keen on merging this version to your repo so I can send a PR.

Here's Popi in RVIZ walking after configuration. Nominal Height set to 0.4
![popi](https://user-images.githubusercontent.com/5070395/95649860-6662cd00-0b12-11eb-9549-64ce3dd1ef12.gif)

**Comment by popi-mkx3** [comment]
Hi grassjelly,

I am deeply sorry it took me so long to reply, I did not see my email notification about your answers !
I just tried it out and had a look at the changes you made on the xacro file. I have to admit I am a bit ashamed for not figuring out how to make these changes myself, thank you for teaching me.

I had to make some adjustments though, especially I had to remove the <hardwareInterface> tag generated in every <actuator> tags because it notified me with an error when trying to use the bringup.launch file. I also had to replace the boxes dimensions because some of them had negative values.
By the way, thank you for replacing the meshes with boxes, I know it's better for performances but never took the time to do it.


At this point, it worked really well in RViz, but the legs would stay stretched out in Gazebo. After some times, I figured out what the problem was. The knee couldn't bend because of the limits. I changed :
lower = "-0.26" upper = "1.92" into lower = "-1.92" upper = "0.26" 
And this works better. Though I had to tune the gaits parameters to find something that would move the intended way. With a trials & errors method, I came up with these parameters: 

knee_orientation : ">>"
pantograph_leg : false
odom_scaler: 1.0
max_linear_velocity_x : 0.5
max_linear_velocity_y : 0.25
max_angular_velocity_z : 1.0
com_x_translation : -0.08
swing_height : 0.15
stance_depth : 0
stance_duration : 0.6
nominal_height : 0.487

Also had to up the knees velocity to 3.14 rd/s, though I'm absolutely certain my physical robot couldn't keep up the pace.

If you have better insights to tune these parameters I'll be really happy to have your opinion. In the meantime, I'd be glad to receive your PR. I'll update it with my adjustments afterwards.

I'd be happy to have my humble model join your official repo as well.


Thanks again for your amazing work,
Lucas
